File: 2020/2020W14/preppindataw14.R
library(dplyr)
library(tidyr)
library(readr)
library(scales)

final <- read_csv("E:/PD Week 14 Input.csv", 
                  col_types = cols(Fang_length = col_number(), 
                                   Prosoma_height = col_number(), Prosoma_length = col_number(), 
                                   Prosoma_width = col_number(), Tibia_I_length = col_number(), 
                                   Total_body_length = col_number())) %>%
  drop_na() %>%
  group_by(`Species`) %>%
  filter(n()>=10) %>%
  pivot_longer(., cols=contains('_'), names_to='Trait', values_to='Value') %>%
  group_by(`Species`,`Trait`) %>%
  summarise('Value' = mean(`Value`)) %>%
  group_by(`Trait`) %>%
  mutate('Max Value' = max(`Value`),
         'Min Value' = min(`Value`),
         'Normalised Value' = (`Value`-`Min Value`)/(`Max Value`-`Min Value`)) %>%
  ungroup() %>% 
  mutate('Species' = gsub('_',' ',`Species`),
         'Trait' = gsub('_',' ',`Trait`)) 

# Alternative method: Calculate with scales::rescale
final <- final %>%
  group_by(`Trait`) %>%
  mutate('Normalised Value Alt' = rescale(`Value`, to=c(0,1)))

View(final)

File: 2020/2020W14/PreppindataW14.py
import pandas as pd
import re
from sklearn.preprocessing import minmax_scale

final = pd.read_csv(r'E:/PD Week 14 Input.csv',dtype=str)
for col in [c for c in final.columns if re.search('(\_)',c)]:
    final[col] = pd.to_numeric(final[col], errors='coerce')
final = final.dropna()
final['Species Count'] = final['Code'].groupby(final['Species']).transform('count')
final = final[final['Species Count']>=10].drop(['Code', 'Family', 'Sex', 'Location', 'Species Count'], axis=1).copy()
final = final.melt(id_vars = 'Species', value_name = 'Value', var_name = 'Trait')
final['Species'] = final['Species'].str.replace('_',' ')
final['Trait'] = final['Trait'].str.replace('_',' ')
final = final.groupby(['Species','Trait'], as_index=False).agg({'Value':'mean'})
final['Max Value'] = final['Value'].groupby(final['Trait']).transform('max')
final['Min Value'] = final['Value'].groupby(final['Trait']).transform('min')
final['Normalised Value'] = (final['Value'] - final['Min Value']) / (final['Max Value'] - final['Min Value'])

# Alternative method: Calculate with sklearn.preprocessing.minmax_scale
final['Normalised Value Alt'] = final.groupby('Trait')['Value'].transform(lambda x: minmax_scale(x))
